File: flutter/sample_ui/lib/src/state/use_smileid_sample_user_details_requirement.dart
import 'use_smileid_sample_profiles.dart';
import 'use_smileid_sample_token_decoder.dart';

/// What the consent form still has to collect, which a token can narrow.
class UseSmileIDSampleUserDetailsRequirement {
  /// Everything required, which is what a launch with no token gets.
  const UseSmileIDSampleUserDetailsRequirement({
    this.firstName = true,
    this.lastName = true,
    this.contact = true,
  });

  /// Whether a given name is still needed.
  final bool firstName;

  /// Whether a family name is still needed.
  final bool lastName;

  /// Whether an email OR a phone is still needed; one of, never both.
  final bool contact;

  /// Nothing left to ask, so the form has no reason to appear.
  bool get isSatisfied => !firstName && !lastName && !contact;

  /// Whether [field] is one the token already supplied, which is why it renders as provided.
  bool supplies(UseSmileIDSampleUserField field) => switch (field) {
    UseSmileIDSampleUserField.firstName => !firstName,
    UseSmileIDSampleUserField.lastName => !lastName,
    // Contact is "one of", so a bound email leaves phone askable: only the requirement lifts.
    UseSmileIDSampleUserField.email || UseSmileIDSampleUserField.phone => false,
  };

  @override
  bool operator ==(Object other) =>
      other is UseSmileIDSampleUserDetailsRequirement &&
      other.firstName == firstName &&
      other.lastName == lastName &&
      other.contact == contact;

  @override
  int get hashCode => Object.hash(firstName, lastName, contact);

  /// Whether [details] satisfies what is still outstanding.
  bool isSatisfiedBy(UseSmileIDSampleUserDetails details) =>
      (!firstName || details.firstName.trim().isNotEmpty) &&
      (!lastName || details.lastName.trim().isNotEmpty) &&
      (!contact ||
          details.email.trim().isNotEmpty ||
          details.phone.trim().isNotEmpty);

  /// The row's label, which gains "(optional)" only once a token has covered contact.
  String labelFor(UseSmileIDSampleUserField field) => switch (field) {
    UseSmileIDSampleUserField.firstName => 'First name',
    UseSmileIDSampleUserField.lastName => 'Last name',
    UseSmileIDSampleUserField.email => contact ? 'Email' : 'Email (optional)',
    UseSmileIDSampleUserField.phone => contact ? 'Phone' : 'Phone (optional)',
  };

  /// What this requirement still asks for; the SCREEN decides when nothing is outstanding.
  String get prompt {
    final List<String> outstanding = <String>[
      if (firstName) 'first name',
      if (lastName) 'last name',
      if (contact) 'an email or phone number',
    ];
    if (outstanding.isEmpty) {
      return 'Tap any field to edit.';
    }
    if (outstanding.length == 1) {
      final String only = outstanding.single;
      return '${only[0].toUpperCase()}${only.substring(1)} is required.';
    }
    return 'Required: ${outstanding.join(', ')}.';
  }
}

/// The requirement a token leaves behind: the SDK's union rule, field for field.
UseSmileIDSampleUserDetailsRequirement useSmileIDSampleUserDetailsRequirement(
  UseSmileIDSampleTokenBindings? bindings,
) => UseSmileIDSampleUserDetailsRequirement(
  firstName: bindings?.givenNames != true,
  lastName: bindings?.lastName != true,
  contact: !(bindings?.email == true || bindings?.phoneNumber == true),
);

/// A duplicate of the SDK's internal `bindsRequiredUserDetails`: both names plus one contact field.
extension UseSmileIDSampleRequiredUserDetails on UseSmileIDSampleTokenBindings {
  /// Whether the token binds enough for the SDK to stop requiring `userDetails`.
  bool get bindsRequiredUserDetails =>
      useSmileIDSampleUserDetailsRequirement(this).isSatisfied;
}

File: expo/sample-ui/src/state/use-smile-id-sample-user-details-requirement.ts
import type { UseSmileIDSampleTokenBindings } from './use-smile-id-sample-token-decoder';
import type { UseSmileIDSampleUserDetails } from './use-smile-id-sample-profiles';
import { UseSmileIDSampleUserField, type UseSmileIDSampleUserFieldSpec } from '../model/use-smile-id-sample-user-fields';

export type { UseSmileIDSampleTokenBindings };

/// What the form must still collect: the SDK's rule minus what the token binds.
export type UseSmileIDSampleUserDetailsRequirement = {
  readonly firstName: boolean;
  readonly lastName: boolean;
  /// One of email or phone, which is why neither row is individually supplied.
  readonly contact: boolean;
};

/// No token, so the form asks for everything the SDK's own validator would.
export const smileIDSampleRequirementDefaults: UseSmileIDSampleUserDetailsRequirement = {
  firstName: true,
  lastName: true,
  contact: true,
};

/// The requirement a token leaves behind, which is the only thing that lifts a row.
export const smileIDSampleRequirementFrom = (
  bindings: UseSmileIDSampleTokenBindings | null | undefined,
): UseSmileIDSampleUserDetailsRequirement => ({
  firstName: bindings?.givenNames !== true,
  lastName: bindings?.lastName !== true,
  contact: !(bindings?.email === true || bindings?.phoneNumber === true),
});

/// Nothing left to ask, so the form has no reason to appear.
export const smileIDSampleRequirementSatisfied = (
  requirement: UseSmileIDSampleUserDetailsRequirement,
): boolean => !requirement.firstName && !requirement.lastName && !requirement.contact;

/// No relevant binding at all, the one case the SDK's own validator can still decide.
export const smileIDSampleRequirementBindsNothing = (
  requirement: UseSmileIDSampleUserDetailsRequirement,
): boolean => requirement.firstName && requirement.lastName && requirement.contact;

/// Whether the form has collected what the requirement still asks of it.
export const smileIDSampleDetailsSatisfy = (
  details: UseSmileIDSampleUserDetails,
  requirement: UseSmileIDSampleUserDetailsRequirement,
): boolean =>
  (!requirement.firstName || details.firstName.trim().length > 0) &&
  (!requirement.lastName || details.lastName.trim().length > 0) &&
  (!requirement.contact || details.email.trim().length > 0 || details.phone.trim().length > 0);

/// Whether the token already supplied this row, which is why it renders as provided.
export const smileIDSampleRequirementSupplies = (
  requirement: UseSmileIDSampleUserDetailsRequirement,
  field: UseSmileIDSampleUserField,
): boolean => {
  if (field === UseSmileIDSampleUserField.FirstName) return !requirement.firstName;
  if (field === UseSmileIDSampleUserField.LastName) return !requirement.lastName;
  // The rule is "one of", so a bound email leaves phone askable. Only the requirement itself lifts.
  return false;
};

/// A contact row stops saying "optional" the moment one of the two is actually required.
export const smileIDSampleRequirementLabel = (
  requirement: UseSmileIDSampleUserDetailsRequirement,
  field: UseSmileIDSampleUserFieldSpec,
): string => {
  if (!requirement.contact) return field.label;
  if (field.id === UseSmileIDSampleUserField.Email) return 'Email';
  if (field.id === UseSmileIDSampleUserField.Phone) return 'Phone';
  return field.label;
};

const titlecase = (text: string): string => text.charAt(0).toUpperCase() + text.slice(1);

/// The sentence under the form, which has to name what is actually outstanding.
export const smileIDSampleRequirementPrompt = (
  requirement: UseSmileIDSampleUserDetailsRequirement,
): string => {
  const outstanding: string[] = [];
  if (requirement.firstName) outstanding.push('first name');
  if (requirement.lastName) outstanding.push('last name');
  if (requirement.contact) outstanding.push('an email or phone number');
  if (outstanding.length === 0) return 'Tap any field to edit.';
  if (outstanding.length === 1) return `${titlecase(outstanding[0]!)} is required.`;
  return `Required: ${outstanding.join(', ')}.`;
};
